src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+    glm::vec3 u = getU(normalizedDirection);
+    glm::vec3 v = glm::cross(u, normalizedDirection);
+    float angle = m_effectParams->angle;
+    float phi = getRandom(-1 * glm::pi<float>(), glm::pi<float>());
+    float z = getRandom(cos(angle), 1);
+    float theta = acosf(z);
+    p->velocity = sin(theta) * (cos(phi) * u + sin(phi) * v) + cos(theta) * normalizedDirection;
+}
+
+glm::vec3 ParticleEffect::getU(glm::vec3 directionNorm) {
+    glm::vec3 randVec = glm::ballRand(1.0f);
+    glm::vec3 u = normalize(cross(directionNorm, randVec));
+    while (u == glm::vec3(0.0f)) {
+        randVec = glm::ballRand(1.0f);
+        u = normalize(cross(directionNorm, randVec));
+    }
+    return u;
+}
+
+
+float ParticleEffect::getRandom(float low, float high) {
+    return low + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX / (high - low)));
+}
+
+Vector<ParticleEffect::Particle*> ParticleEffect::generateParticles() {
+    Vector<Particle*> vp = Vector<Particle*>();
+    for (int i = 0; i < m_effectParams->n; i++) {
+        Particle *p = makeParticle(i);
+        vp.push_back(p);
+    }
+    return vp;
+}
+
+// This and the inits need to eventually be cleaned up
+ParticleEffect::Particle* ParticleEffect::makeParticle(int i) {
+    Particle *p = new Particle();
+    
+    initParticle(p, i, 0, 0);
+
+    //Determine if active/inactive
+    if (m_effectParams->rate == 0.0f) {
+        p->active = true;
+    }
+    else {
+        p->active = false;
+    }
+
+    // Get Orientation ID
+    if (m_effectParams->orientations == 0) {
+        p->orientationID = 0;
+    }
+    else {
+        p->orientationID = p->i % m_effectParams->orientations;
+    }
+    
+
+    initPosition(p);
+
+    initVelocity(p);
+
+    return p;
+}
+
+void ParticleEffect::initParticle(Particle *p, int i, int meshID, int modelTextureID) {
+    p->i = i;
+    p->meshID = meshID;
+    p->modelTextureID = modelTextureID;
+    p->life = 0.0f;
+}
+
+void ParticleEffect::initVelocity(Particle *p) {
+    switch (m_effectParams->type) {
+        case ParticleEffect::SPHERE : {
+            sphereMotion(p);
+            break;
+        }
+        case ParticleEffect::DISK : {
+            diskMotion(p);
+            break;
+        }
+        case ParticleEffect::CONE: {
+            coneMotion(p);
+            break;
+        }
+    }
+
+    //Add random variance if specified
+    if (m_effectParams->variance != 0.0f) {
+        glm::vec3 randVec = m_effectParams->variance * normalize(glm::ballRand(1.0f));
+        p->velocity += randVec;
+    }
+
+    if (m_effectParams->randomDistribution) {
+        p->velocity *= getRandom(0.25f, 1.0f);
+    }
+
+    //set speed
+    p->velocity *= m_effectParams->magnitude;
+}
+
+void ParticleEffect::initPosition(Particle *p) {
+    p->position = m_anchor + m_offset + m_life * m_velocity;
+}
+
+Vector<glm::vec3> * ParticleEffect::getActiveParticlePositions() {
+    Vector<glm::vec3> *positions = new Vector<glm::vec3>();
```

Comment:
Is this being deleted anywhere?

Reply:
no

Reply:
see other comments about not using vectors this way